docs/dogstatsd/internals.md

```diff
@@ -0,0 +1,53 @@
+![Dogstatsd Internals Schema](internals_schema.png)
+
+## Packet
+
+In Dogstatsd, a Packet is a bytes array containing one or multiple metrics in the Dogstatsd format (separated by a \n when there are several). Its maximum size is dogstatsd_buffer_size.
+
+## PacketAssembler
+
+Input: a datagram from an UDP socket
+Output: a Packet containing multiple metrics packed together, separated by a \n
+
+The PacketAssembler is gathering multiple datagrams into one Packet of maximum size dogstatsd_buffer_size sending it to the PacketsBuffer, avoiding to run the whole parsing pipeline with only 1 metric per packet. The bytes buffer used comes from the PacketPool, avoiding having to re-allocate bytes buffer every time. Please note that the UDS pipeline does not use the PacketAssembler because each UDS packet also contains metadata (origin tags) which are used to enrich the metrics tags, making them impossible to be packed together by the PacketAssembler.
+
+The PacketAssembler is not allocating a bytes array every time it has to use one, it is getting it from a pool containing pre-allocated ones. The pool is never empty, it takes care of allocating a new bytes array when it’s needed.
+Once this bytes array has been fully assembled by the PacketAssembler, it is sent to the rest of the Dogstatsd pipeline and its ownership is transferred to each part using it (PacketsBuffer, Worker). Eventually, the Worker takes care of returning it to the pool when it has processed its content.
+
+## PacketBuffer
+
+Input: a Packet containing one or several metrics in the Dogstatsd format (separated by a \n)
+Output: multiple Packets send in a row to the Worker
+
+The PacketsBuffer is buffering multiple Packets (in a slice), this way the parsing part of the pipeline is going through several Packets in a row instead of only one each time it is called. This leads to less CPU usage. PacketsBuffer sends the Packets for processing when either:
```

Comment:
```suggestion
The PacketsBuffer buffers multiple Packets (in a slice), this way the parsing part of the pipeline is going through several Packets in a row instead of only one each time it is called. This leads to less CPU usage. PacketsBuffer sends the Packets for processing when either:
```